File: 백준/Gold/1238. 파티/파티.kt
import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.util.PriorityQueue

val INF : Int = 987654321;

fun main(args: Array<String>) {

    val br = BufferedReader(InputStreamReader(System.`in`));
    val bw = BufferedWriter(OutputStreamWriter(System.out));

    var num = br.readLine().split(" ");
    val N = num[0].toInt();
    val M = num[1].toInt();
    val X = num[2].toInt();

    val adj = MutableList(N+2) { mutableListOf<Pair<Int,Int>>() };
    val retDist = IntArray(N+2) {init -> 0};

    (1..M).forEach() {
        num = br.readLine()!!.split(" ");
        adj[num[0].toInt()].add(Pair(num[2].toInt(),num[1].toInt()));
    }

    // it마을에서 X마을 까지의 최단거리 구하기
    (1..N).forEach() {
        val pq = PriorityQueue<Pair<Int,Int>>(compareBy { it.first });
        val dist = IntArray(N+1) {init -> INF};

        pq.add(Pair(0,it));
        dist[it] = 0;

        while(pq.isNotEmpty()) {
            val here = pq.peek().second;
            val here_dist = pq.poll().first;

            if(dist[here] != here_dist) continue;

            for(i in 0..adj[here].lastIndex) {
                val _there = adj[here][i].second;
                val _there_dist = adj[here][i].first;
                if(dist[_there] > dist[here] + _there_dist) {
                    dist[_there] = dist[here] + _there_dist;
                    pq.add(Pair(dist[_there],_there));
                }
            }
        }

        // 최단거리
        if(it==X) {
            for(i in 1..N) {
                retDist[i] += dist[i];
            }
        }
        else {
            retDist[it] += dist[X];
        }

    }

    var ret = 0;
    var mx = 0;
    (1..N).forEach() {
        //bw.write(retDist[it].toString() + " ");
        if(it != X && mx < retDist[it]) {
            ret = it;
            mx = retDist[it];
        }
    }

    println(mx);

    br.close();
    bw.flush();
    bw.close();

}

File: 백준/Gold/1238. 파티/파티.cc
#include <bits/stdc++.h>
using namespace std;

int N, M, X, u, v, w, ret;
int dist[100004], retDist[100004];
const int INF = 1e9+7;
vector<pair<int,int>> adj[100004];

int main(void) {
    ios_base::sync_with_stdio(0); cin.tie(0); cout.tie(0);

    cin >> N >> M >> X;

    for(int i = 0; i < M; i++) {
        cin >> u >> v >> w;
        adj[u].push_back(make_pair(w,v));
    }
    
    for(int startV = 1; startV <= N; startV++) {
        fill(&dist[0], &dist[100003], INF);
        priority_queue<pair<int,int>,vector<pair<int,int>>,greater<pair<int,int>>> pq;

        pq.push(make_pair(0,startV));
        dist[startV] = 0;

        while(pq.size()) {
            int here = pq.top().second;
            int here_dist = pq.top().first;
            pq.pop();

            if(dist[here] != here_dist) continue;

            for(auto it : adj[here]) {
                int there = it.second;
                int there_dist = it.first;
                if(dist[there] > dist[here]+there_dist) {
                    dist[there] = dist[here]+there_dist;
                    pq.push(make_pair(dist[there],there));
                }
            }
        }

        if(startV == X) {
            for(int i = 1; i <= N; i++) retDist[i] += dist[i];
        }
        else retDist[startV] += dist[X];
    }

    for(int i = 1; i <= N; i++) {
        ret = max(ret,retDist[i]);
    }

    cout << ret;
}